monai/data/dataset.py

```diff
@@ -190,35 +199,149 @@ def _pre_first_random_cachecheck(self, item_transformed):
+            db_name: lmdb database file name. Defaults to "monai_cache".
+            pickle_protocol: pickle protocol version. Defaults to pickle.HIGHEST_PROTOCOL.
+                https://docs.python.org/3/library/pickle.html#pickle-protocols
+            lmdb_kwargs: additional keyword arguments to the lmdb environment.
+                for more details please visit: https://lmdb.readthedocs.io/en/release/#environment-class
+        """
+        super().__init__(data=data, transform=transform, cache_dir=cache_dir, hash_func=hash_func)
+        if not self.cache_dir:
+            raise ValueError("cache_dir must be specified.")
+        self.db_file = self.cache_dir / f"{db_name}.lmdb"
+        self.pickle_protocol = pickle_protocol
+        self.lmdb_kwargs = lmdb_kwargs or {}
+        if not self.lmdb_kwargs.get("map_size", 0):
+            self.lmdb_kwargs["map_size"] = 1024 ** 4  # default map_size
+
+        # create cache
+        self._fill_cache()
+        # read-only database env
+        self.lmdb_kwargs["readonly"] = True
+        if self.lmdb_kwargs.get("lock", None) is None:
+            self.lmdb_kwargs["lock"] = False
+        self._read_env = lmdb.open(path=f"{self.db_file}", subdir=False, **self.lmdb_kwargs)
+
+    def _fill_cache(self):
+        print(f"Accessing lmdb file: {self.db_file.absolute()}.")
+        self.lmdb_kwargs["readonly"] = False
+        env = lmdb.open(path=f"{self.db_file}", subdir=False, **self.lmdb_kwargs)
+        if not has_tqdm:
+            warnings.warn("LMDBDataset: tqdm is not installed. not displaying the caching progress.")
+        for item in tqdm(self.data) if has_tqdm else self.data:
+            key = self.hash_func(item)
+            done, retry, val = False, 5, None
+            while not done and retry > 0:
+                try:
+                    with env.begin(write=True) as txn:
+                        with txn.cursor() as cursor:
+                            done = cursor.set_key(key)
+                            if done:
+                                continue
+                        if val is None:
+                            val = self._pre_first_random_transform(deepcopy(item))  # keep the original hashed
+                            val = pickle.dumps(val, protocol=self.pickle_protocol)
+                        txn.put(key, val)
+                    done = True
+                except lmdb.MapFullError:
+                    done, retry = False, retry - 1
+                    size = env.info()["map_size"]
+                    new_size = size * 2
+                    print(f"Resizing the cache database from {int(size) >> 20}MB to {int(new_size) >> 20}MB.")
+                    env.set_mapsize(new_size)
+            if not done:  # still has the map full error
+                size = env.info()["map_size"]
+                env.close()
+                raise ValueError(f"Please increase the map size of lmdb (current size={size}).")
```

Comment:
The exception message should say that the LMDB map size was reached and that it needs to be increased, so something like `f"LMBD map size reached, increase size above current size of {size}"`